crates/core/tedge_actors/src/servers/requests.rs

```diff
@@ -0,0 +1,229 @@
+                let _ = reply_to.send(result).await;
+            });
+
+            // Send the response back to the client
+            self.messages.send_response_once_done(pending_result)
+        }
+
+        Ok(())
+    }
+}
+
+/// A message box for services that handles requests concurrently
+pub struct ConcurrentServerMessageBox<Request: Debug, Response> {
+    /// Max concurrent requests
+    max_concurrency: usize,
+
+    /// Message box to interact with clients of this service
+    requests: LoggingReceiver<RequestEnvelope<Request, Response>>,
+
+    /// Pending responses
+    pending_responses: futures::stream::FuturesUnordered<PendingResult>,
+}
+
+type PendingResult = tokio::task::JoinHandle<()>;
+
+impl<Request: Message, Response: Message> ConcurrentServerMessageBox<Request, Response> {
+    pub(crate) fn new(
+        max_concurrency: usize,
+        requests: LoggingReceiver<RequestEnvelope<Request, Response>>,
+    ) -> Self {
+        ConcurrentServerMessageBox {
+            max_concurrency,
+            requests,
+            pending_responses: futures::stream::FuturesUnordered::new(),
+        }
+    }
+
+    async fn next_request(&mut self) -> Option<RequestEnvelope<Request, Response>> {
+        if self.await_idle_processor().await.is_break() {
+            return None;
+        }
+
+        loop {
+            tokio::select! {
+                Some(request) = self.requests.recv() => {
+                    return Some(request);
+                }
+                Some(result) = self.pending_responses.next() => {
+                    if let Err(err) = result {
+                        error!("Request failed with: {err}");
+                    }
+                }
+                else => {
+                    return None
+                }
+            }
+        }
+    }
+
+    async fn await_idle_processor(&mut self) -> ControlFlow<(), ()> {
+        if self.pending_responses.len() < self.max_concurrency {
+            return ControlFlow::Continue(());
+        }
+
+        tokio::select! {
+            Some(result) = self.pending_responses.next() => {
+                if let Err(err) = result {
+                    error!("Request failed with: {err}");
+                }
+                ControlFlow::Continue(())
+            },
+            // recv consumes the message from the channel, so we can't just use
+            // a regular return, because then next_request wouldn't see it
+            //
+            // a better approach would be to do select on top-level entry point,
+            // then we'd be sure we're able to cancel when anything happens, not
+            // just when waiting for pending_responses.
+            Some(RuntimeRequest::Shutdown) = self.requests.recv_signal() => {
+                ControlFlow::Break(())
+            }
+            else => ControlFlow::Break(())
+        }
+    }
+
+    pub fn send_response_once_done(&mut self, pending_result: PendingResult) {
```

Comment:
```suggestion
    pub fn register_request_handler(&mut self, pending_result: PendingResult) {
```

Reply:
I avoided to change things when not required, to ease code reviews. I can finalize this PR with commits addressing such clean-up.

That said, I have a preference for `send_response_once_done` (what / purpose) over `register_request_handler` (how / mechanism).